### PR TITLE
Create `mach bootstrap` based on Mozilla's mozboot bootstrapper

### PR DESCRIPTION
**Pull request by UK992**
Fixes https://github.com/servo/servo/issues/12914
I've made this few weeks ago, its an example how could everything looks like. 
It downloads and setup all needed dependencies for MSVC.
It's has version in case if some dependencies need to be updated.
Zip files and folder in zip need to be named `<dep>-<version>`.
Also if cmake already exist in PATH, it won't download it again.

I want opinion on that, if this is right approaches and how to improve it.

cc @vvuk

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12916)

### CODE REVIEW

**Comment by SaschaNaz** [comment]
Can the command window be kept when error occurred? Users won't see the error messages unless they manually run `cmd`.

**Comment by UK992** [comment]
@SaschaNaz I don't understand, can you please show which part of code do you mean?

**Comment by SaschaNaz** [comment]
@UK992 I was talking about mach.bat ECHO lines but never mind, the error message was thrown from mach rather than mach.bat. Is there a way to keep cmd window when mach throws any error?

File: python/servo/bootstrap_commands.py [hunk 1]
```diff
@@ -126,6 +131,79 @@ def env(self):
...
+        for dep in deps:
+            dep_split = dep.split('=')
+            dep_name = dep_split[0]
+            dep_version = dep_split[1]
+            dep_full_name = dep_name + " " + dep_version
+
+            # Don't download CMake if already exists in PATH
+            if dep_name == "cmake":
+                if spawn.find_executable(dep_name):
+                    continue
```
Comment (Wafflespeanut) [inline python/servo/bootstrap_commands.py]: This could be moved out of the loop?

**Comment by UK992** [comment]
In last commit i created ``mach bootstrap`` command based on [Mozilla's mozboot](https://dxr.mozilla.org/mozilla-central/source/python/mozboot/mozboot), still needs some cleanups but its for example how could servo bootstrapper looks like.
And can be extended to other platforms, since it's most work already done.

File: python/servo/bootstrap_commands.py (hunk 1 above)
Comment (UK992) [inline python/servo/bootstrap_commands.py]: Hardly, considering that it compares with the names in the list.

File: python/servo/bootstrapper/base.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    def install_mobile_android_packages(self):
+        '''
+        Install packages required to build Servo for Android.
+        '''
+        raise NotImplementedError('Cannot bootstrap Servo for Android: '
+                                  '%s does not yet implement install_mobile_android_packages()'
+                                  % __name__)
+
+    def which(self, name):
+        """Python implementation of which.
```
Comment (Wafflespeanut) [inline python/servo/bootstrapper/base.py]: Can't we just use `distutils.spawn.find_executable`? (like we do in `mach_bootstrap.py`)

**Comment by Wafflespeanut** [comment]
I don't have much ideas/suggestions regarding MSVC bootstrapping (apart from python'ish style nits, which I'll address in the end). Pinging @larsbergstrom @vvuk and @metajack to look into this sometime...

File: python/servo/bootstrapper/base.py (hunk 2 above)
Comment (UK992) [inline python/servo/bootstrapper/base.py]: done

File: appveyor.yml [hunk 3]
```diff
@@ -45,26 +45,26 @@ cache:
...
-           $env:PATH="$pwd\ninja;$env:PATH" ;
-           $env:OPENSSL_INCLUDE_DIR="$pwd\openssl-bin\openssl-1.0.1t-vs2015\include" ;
-           $env:OPENSSL_LIB_DIR="$pwd\openssl-bin\openssl-1.0.1t-vs2015\lib64" ;
-           $env:OPENSSL_LIBS="ssleay32MD:libeay32MD" ;
-           $env:FFMPEG_INCLUDE_DIR="$pwd\ffmpeg-bin\include" ;
-           $env:FFMPEG_LIB_DIR="$pwd\ffmpeg-bin\lib" ;
-           $env:FFMPEG_LIBS="avformat:avcodec:avutil" ;
-        }'
-  - if %BUILD_ENV%==msvc call "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\amd64\vcvars64.bat"
+#  - ps: 'if ($env:BUILD_ENV -eq "msvc") {
```
Comment (larsbergstrom) [inline appveyor.yml]: Can you change this to instead do `./mach bootstrap-msvc` and see if that works on AppVeyor?

File: python/servo/bootstrap_commands.py [hunk 4]
```diff
@@ -126,6 +131,79 @@ def env(self):
...
+    @CommandArgument('--force', '-f',
+                     action='store_true',
+                     help='Force download even if a copy already exists')
+    def bootstrap_msvc(self, force=False):
+        # Don't download MSVC dependencies on non-Windows platform
+        if not is_windows():
+            return print("Unsupported platform.")
+
+        msvc_deps_dir = path.join(self.context.sharedir, "msvc-dependencies")
+        msvc_deps_url = "https://dl.dropboxusercontent.com/u/25971865/msvc-deps/"
```
Comment (larsbergstrom) [inline python/servo/bootstrap_commands.py]: We will definitely want to move these to one of the Servo S3 buckets. I like that you've versioned the files, which is something we've definitely found a need for in the past.

File: appveyor.yml (hunk 3 above)
Comment (UK992) [inline appveyor.yml]: ``bootstrap-msvc`` is replaced by ``bootstrap`` in newer commit, which also support gnu builds, but for msvc build, command is being run in ``ensure_bootstrapped``. But i did that for gnu builds.

**Comment by larsbergstrom** [comment]
@UK992 I love this work - thanks a ton for doing it! Just two more fixes:
1) Change the URL from your dropbox folder to, e.g., use `https://servo-rust.s3.amazonaws.com/msvc-deps/moztools-0.0.1-5.zip`. I've copied all the dependencies up there.
2) Squash the fixup in, as appropriate.

@bors-servo delegate=uk992

**Comment by UK992** [comment]
@larsbergstrom done!
@Wafflespeanut Can you take a look again for python'ish style nits?

**Comment by UK992** [comment]
I've made some renamings:
check_installed_packages -> ensure_system_packages
install_packages -> install_system_packages

**Comment by Wafflespeanut** [comment]
I just had a brief overview on this, and I'm cool with most of the stuff. Let's land this thing happily and if there's something dirty, we'll cleanup later :smile:

**Pull request merged**